### PR TITLE
feat: Support Not Requiring `projectId` When Not Required

### LINKED ISSUE #755: Requires projectId even when not used
#### Environment details

  - OS: linux
  - Node.js version: v10.15.3
  - npm version: 6.4.1
  - `google-auth-library` version: 5.0.0

#### Steps to reproduce

  1. Run `gcloud auth revoke` (if needed)
  2. Run `gcloud auth application-default login`
  3. run script
```
const {GoogleAuth} = require('google-auth-library')

async function run () {
  const auth = new GoogleAuth({
    scopes: 'https://www.googleapis.com/auth/cloud-platform'
  });
  client = await auth.getClient();
  console.info(client)
}
run();
```
  4. Get error
Error: Unable to detect a Project Id in the current environment. 
To learn more about authentication and Google APIs, visit: 
https://cloud.google.com/docs/authentication/getting-started
    at _getDefaultProjectIdPromise.Promise (/usr/local/google/home/kaylanguyen/Downloads/buildartifacts-npm-tools/node_modules/google-auth-library/build/src/auth/googleauth.js:90:31)
    at process._tickCallback (internal/process/next_tick.js:68:7)

#### Expectation:
- Able to create a client and get an access token without having to clarify projectID because I don't need it.

#### Work around:
- Have a fake projectId
```
const auth = new GoogleAuth({
        scopes: 'https://www.googleapis.com/auth/cloud-platform',
        projectId: 'fake-project'
    });
```

Other gotchas:
- export GOOGLE_APPLICATION_CREDENTIALS=<path/to/service/account> works
- Runing both `gcloud auth login` and `gcloud auth application-default login` then the script works

### PR DESCRIPTION
**Pull request by danielbankhead**
There are numerous cases where `projectId` isn't required to use `GoogleAuth`.

In fact, I'm failing to find a place where `projectId` is a hard requirement within the library (including `sign`, which erroneously requires it and will be removed in this pr: https://github.com/googleapis/google-auth-library-nodejs/pull/1447).

Fixes #755 🦕

### CODE REVIEW

**Comment by danielbankhead** [comment]
This PR should be followed-up with an issue for removing `#getProjectIdOptional` in favor of fixing & using `getProjectId` (with the corrected `Promise<string | null>`) in a future major release.

File: src/auth/googleauth.ts [hunk 1]
```diff
@@ -191,9 +198,32 @@ export class GoogleAuth<T extends AuthClient = JSONClient> {
...
+      if (
+        e instanceof Error &&
+        e.message === GoogleAuthExceptionMessages.NO_PROJECT_ID_FOUND
+      ) {
+        return null;
+      } else {
+        throw e;
+      }
+    }
+  }
```
Comment (ruyadorno) [inline src/auth/googleauth.ts:222]: Leveraging the Promises API (similar to the previous PR) 😄 might simplify it a bit here again:

```suggestion
  #getProjectIdOptional(): Promise<string | null> {
    return this.getProjectId()
      .catch (e => {
        if (e.message === GoogleAuthExceptionMessages.NO_PROJECT_ID_FOUND) {
          return null;
        } else {
          throw e;
        }
      })
  }
```

**Review by ruyadorno — COMMENTED** [review]
LGTM 👍 just a small suggestion

File: src/auth/googleauth.ts [hunk 2]
```diff
@@ -191,9 +198,32 @@ export class GoogleAuth<T extends AuthClient = JSONClient> {
...
+   *
+   * @returns Promise that resolves with project id (or `null`)
+   */
+  async #getProjectIdOptional(): Promise<string | null> {
+    try {
+      return await this.getProjectId();
+    } catch (e) {
+      if (
+        e instanceof Error &&
+        e.message === GoogleAuthExceptionMessages.NO_PROJECT_ID_FOUND
```
[the remark this answers is not in the document]
Comment (danielbankhead) [reply src/auth/googleauth.ts:215]: Understandable; the `instanceof Error` check is more-so a type assertion. By default, TS considers 'e' `unknown`

**Review by ruyadorno — APPROVED** [review]

**Pull request merged**